spokestack/pipeline.py

```diff
@@ -0,0 +1,131 @@
+"""
+This module contains the speech pipeline which manages the components
+for processing speech.
+"""
+
+from spokestack.context import SpeechContext
+
+
+class SpeechPipeline:
+    """ Pipeline for managing speech components.
+
+    Args:
+        input_source: source of audio input
+        stages: components desired in the pipeline
+        **kwargs: additional keyword arguments
+    """
+
+    def __init__(self, input_source, stages, **kwargs) -> None:
+        self._context = SpeechContext(**kwargs)
+        self._input_source = input_source
+        self._stages: list = stages
+        self._is_running = False
+        self._is_managed = False
+        self._is_paused = False
+
+    def _dispatch(self) -> None:
+        is_managed = self._context.is_managed
+        if self._is_managed and not is_managed:
+            for stage in self._stages:
+                stage.reset()
+        self._is_managed = is_managed
+
+        if self._is_paused:
+            pass
+        else:
+            frame = self._input_source.read()
+            for stage in self._stages:
+                if not self._is_managed:
+                    stage(self._context, frame)
+
+    def close(self) -> None:
+        """ Closes the running pipeline """
+        self.stop()
+
+    def activate(self) -> None:
+        """ Activates the pipeline """
+        self._context.is_active = True
+
+    def deactivate(self) -> None:
+        """ Deactivates the pipeline """
+        self._context.is_active = False
+
+    def start(self) -> None:
+        """ Starts input source of the pipeline """
+        if not self._is_running:
+            self._input_source.start()
+            self._is_running = True
+
+    def stop(self) -> None:
+        """ Halts the pipeline """
+        if self._is_running:
+            self._is_running = False
+
+    def pause(self) -> None:
+        """ Stops audio input until resume is called """
+        self._is_paused = True
+        self._is_running = False
+
+    def resume(self) -> None:
+        """ Resumes audio input after a pause """
+        self._is_paused = False
+        self._is_running = True
+
+    def run(self) -> None:
+        """ Runs the pipeline to process speech and cleans up after stop is called """
+        while self._is_running:
```

Comment:
in the case of a synchronous model, where the application calls run on the main thread pauses the pipeline when it receives a recognition event,
`self._is_running` will be set to false when `pause` is called, which will cause the application to exit
it seems that you wouldn't want to set `self._is_running` to false when the pipeline is paused